packages/components/src/color-picker/inputs.js

```diff
@@ -97,63 +107,115 @@ export class Inputs extends Component {
 
 	toggleViews() {
 		if ( this.state.view === 'hex' ) {
-			this.setState( { view: 'rgb' } );
+			this.setState( { view: 'rgb' }, this.resetDraftValues );
 
 			speak( __( 'RGB mode active' ) );
 		} else if ( this.state.view === 'rgb' ) {
-			this.setState( { view: 'hsl' } );
+			this.setState( { view: 'hsl' }, this.resetDraftValues );
 
 			speak( __( 'Hue/saturation/lightness mode active' ) );
 		} else if ( this.state.view === 'hsl' ) {
 			if ( this.props.hsl.a === 1 ) {
-				this.setState( { view: 'hex' } );
+				this.setState( { view: 'hex' }, this.resetDraftValues );
 
 				speak( __( 'Hex color mode active' ) );
 			} else {
-				this.setState( { view: 'rgb' } );
+				this.setState( { view: 'rgb' }, this.resetDraftValues );
 
 				speak( __( 'RGB mode active' ) );
 			}
 		}
 	}
 
-	handleChange( data ) {
-		if ( data.hex ) {
-			if ( isValidHex( data.hex ) ) {
-				this.props.onChange( {
-					hex: data.hex,
-					source: 'hex',
-				} );
-			}
-		} else if ( data.r || data.g || data.b ) {
-			this.props.onChange( {
-				r: data.r || this.props.rgb.r,
-				g: data.g || this.props.rgb.g,
-				b: data.b || this.props.rgb.b,
-				source: 'rgb',
-			} );
-		} else if ( data.a ) {
-			if ( data.a < 0 ) {
-				data.a = 0;
-			} else if ( data.a > 1 ) {
-				data.a = 1;
-			}
+	resetDraftValues() {
+		return this.props.onChange( {
+			state: 'reset',
+		} );
+	}
 
-			this.props.onChange( {
-				h: this.props.hsl.h,
-				s: this.props.hsl.s,
-				l: this.props.hsl.l,
-				a: Math.round( data.a * 100 ) / 100,
-				source: 'rgb',
-			} );
-		} else if ( data.h || data.s || data.l ) {
-			this.props.onChange( {
-				h: data.h || this.props.hsl.h,
-				s: data.s || this.props.hsl.s,
-				l: data.l || this.props.hsl.l,
-				source: 'hsl',
-			} );
+	handleHEXChange( data ) {
+		this.props.onChange( {
+			hex: data.hex,
+			source: 'hex',
+			state: data.state,
+		} );
+	}
+
+	handleRChange( data ) {
```

Comment:
Can we use a higher-order function to auto-generate these functions ( handleRGB/HSL changes ) and avoid duplicate code?
e.g:
```
handleChange( source, component ) {
	return function ( data ) {
		this.props.onChange( {
			...this.props[source],
			sourc,
			state: data.state,
			[component]: data[component]
		} );
	}
}
handleGChange = handleChange( 'rgb', 'g');
```

Reply:
I thought about that and it's a smart use of object destructuring. Thanks for sharing! My concern is that it compresses the code to the point it becomes difficult to reason about: it's difficult to see the object shape for each handler and that `data.r` will override `props.rgb.r` for the R handler in RGB. I lean towards optimizing code for readers. Is this a blocker for you?

Reply:
Also: we can't use that for `handleAChange`.

Reply:
```js
handleChange( source, key, data ) {
	const value = ( NORMALIZERS[ key ] || identity )( data[ key ] );
	this.props.onChange( {
		...this.props[ source ],
		[ key ]: value,
		source,
		state: data.state,
	} );
}
```

with

```js
const NORMALIZERS = {
	a: ( value ) =>
		Math.round( Math.max( 0, Math.min( 1, value ) ) * 100 ) / 100,
};
```

This could naturally be adapted to key not just on `key` but on `( source, key )`, if need be.

Then:

```js
onChange={ this.handleChange.bind( this, 'rgb', 'a' ) }
```

Reply:
>  I lean towards optimizing code for readers. I don't feel strongly about this and I'm open to change. Is this a blocker for you?

I agree we need to optimise for readability. I think that feeling the need to duplicate and specialise in order to improve readability is a code smell, though. The generalisation—`handleChange` or whatever we use—needs to be clearer.

Reply:
@jorgefilipecosta @mcsf Thanks for the push to simplify. I've pushed a new version that hopefully offers a better code compression without compromising readability.